### PR TITLE
Add water heater mode cluster xml

### LINKED ISSUE #34252: Add WaterHeater mode XML
We need to add WaterHeater Mode XML

### PR DESCRIPTION
**Pull request by PeterC1965**
Add WaterHeater mode XML support (issue Fixes #34252)

The spec for the water heater mode cluster can be found at https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/82b2966af17528ab88308ae2341142300ad8a3cb/src/app_clusters/Mode_WaterHeater.adoc

This PR will form a series of PRs that will:

- add sdk support for the water-heater-management and water-heater-mode clusters
- add a water-heater-management example app
- add tests for the water-heater-management and water-heater-mode clusters

### CODE REVIEW

**Review by jamesharrow — APPROVED** [review]
Looks good - but could remove the Water Heater Management bits from this PR.

File: src/app/zap-templates/zcl/zcl.json [hunk 1]
```diff
@@ -641,6 +642,16 @@
         "Power Topology": ["FeatureMap"],
         "Valve Configuration and Control": ["RemainingDuration"],
         "Boolean State Configuration": ["CurrentSensitivityLevel"],
+        "Water Heater Management": [
+            "HeaterTypes",
```
Comment (jamesharrow) [inline src/app/zap-templates/zcl/zcl.json:646]: This lines should be removed from this PR.

File: src/app/zap-templates/zcl/zcl-with-test-extensions.json [hunk 2]
```diff
@@ -643,6 +644,16 @@
         "Power Topology": ["FeatureMap"],
         "Valve Configuration and Control": ["RemainingDuration"],
         "Boolean State Configuration": ["CurrentSensitivityLevel"],
+        "Water Heater Management": [
+            "HeaterTypes",
+            "HeatDemand",
```
[the remark this answers is not in the document]
Comment (PeterC1965) [reply src/app/zap-templates/zcl/zcl-with-test-extensions.json]: Done

File: src/app/zap-templates/zcl/zcl.json (hunk 1 above)
Comment (PeterC1965) [reply src/app/zap-templates/zcl/zcl.json]: Done

File: src/app/zap-templates/zcl/data-model/chip/mode-base-cluster.xml [hunk 3]
```diff
@@ -32,6 +32,7 @@ This is because zap does not currently support generating code for clusters that
     <cluster code="0x005E"/> <!-- Microwave Oven Mode -->
     <cluster code="0x0049"/> <!-- Oven Mode -->
     <cluster code="0x009D"/> <!-- Energy EVSE Mode -->
+    <cluster code="0x009E"/> <!-- Water Heater Management Mode -->
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/mode-base-cluster.xml:35]: That's not what the cluster is called.

File: src/app/zap-templates/zcl/data-model/chip/mode-base-cluster.xml [hunk 4]
```diff
@@ -47,6 +48,7 @@ This is because zap does not currently support generating code for clusters that
     <cluster code="0x005E"/> <!-- Microwave Oven Mode -->
     <cluster code="0x0049"/> <!-- Oven Mode -->
     <cluster code="0x009D"/> <!-- Energy EVSE Mode -->
+    <cluster code="0x009E"/> <!-- Water Heater Managment Mode -->
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/mode-base-cluster.xml:51]: That's not what the cluster is called.

**Review by bzbarsky-apple — APPROVED** [review]

File: src/app/zap-templates/zcl/data-model/chip/mode-base-cluster.xml (hunk 3 above)
Comment (PeterC1965) [reply src/app/zap-templates/zcl/data-model/chip/mode-base-cluster.xml]: Oops/fixed.

File: src/app/zap-templates/zcl/data-model/chip/mode-base-cluster.xml (hunk 4 above)
Comment (PeterC1965) [reply src/app/zap-templates/zcl/data-model/chip/mode-base-cluster.xml]: Oops/fixed.

**Review by andy31415 — APPROVED** [review]

**Pull request merged**